### ISSUE TITLE
[ENH] Fractional differencing and ARFIMA

### ISSUE THREAD

**Issue by j-adamczyk** [sktime/sktime#6700]
**Is your feature request related to a problem? Please describe.**

ARFIMA and various models for fractionally differenced models are useful in modelling low signal-to-noise ratio time series. Implementing fractional differentiation would be a useful addition to sktime.

**Describe the solution you'd like**
Implemented fractional differencing, either as an expansion to the `Differencer` class, or as a separate class. This basically reduces to copying the deprecated [fracdiff](https://github.com/fracdiff/fracdiff) library code. It has scikit-learn compatible transformer, so adding it to sktime should be quite easy.

**Describe alternatives you've considered**
I used to use `fracdiff` library, but now it's unmaintained.

**Comment by fkiraly** [comment]
Hm, this is an interesting one, since it has a complex starting situation:

* existing package unmaintained, but permissive license
* sklearn transformer for fractional differentiation
* "find the lowest fracdiff order" estimator and time series transformer
* ARFIMA, which is not itself contained in `fracdiff`

We could of course implement any of these from scratch, but another option might be easier:

* fork `fracdiff` into `libs` and maintain it there, similar to `pykalman`. The license would allow this easily, the only change we may have to make are compatibility updates to python 3.11-12, and isolating the `statsmodels` dependency.
* write `sktime` transformations for fractional differencing and fracdiff with minimal degree, using the package; optionally, parameter estimator
* ARFIMA would have to be written from scratch, but we may be able to use the `fracdiff` routines once the above is done

@j-adamczyk, do you know anything about the status of `fracdiff`, e.g., are maintainers still reachable, and possibly open to help with a code transfer and fixing compatibility bugs? FYI @simaki

**Comment by fkiraly** [comment]
ok, this is odd - does the primary maintainer simaki no longer have a GitHub account? Any informations on what happened there?

Also, would you be interested to contribute with one of the two ways to proceed, @j-adamczyk?
If we go with option 2 - fork `fracdiff` as an `sktime` library - users can continue (mostly) with the interfaces they are used to, if they do not want to use the `sktime` interfaces.

**Comment by j-adamczyk** [comment]
@fkiraly I don't know anything about the maintainer, I'm quite puzzled about this myself. There was also no notice about looking for maintainers, and nothing in the `fracdiff` issues.

I think that just rewriting this from scratch would be the easiest approach. The main functionality is really the `fdiff` function, which is about 20-30 lines of code, and can easily fit as a class method. Automated differentiation is trivial to implement. `sktime` is more or less standard in terms of interface, and having a separate class will be enough IMO.

Probably will work best as 2 classes:
- `FractionalDifferencer`, to do fractional differentiation, with "auto" option for determining fractional order `d`
- `ARFIMA`, for implementing ARFIMA model (this may be another issue / PR in the future, though)

In the automated option, I would change the default test from ADF to KPSS, since it's more standard and gives better results in my experience. Having this as a single class with regular fractional differentiation should be easy to use.

ARFIMA is then just ARIMA (any implementation, really) with `d=0`. AutoARFIMA can be implemented identically.

I'll try to find time to implement this, but I'm not sure how soon I can start. If anyone else would like to pick this up, feel free to do so.

**Comment by fkiraly** [comment]
ok, makes sense - some comments:

> The main functionality is really the `fdiff` function

We could also create a partial fork then - only the functions we need.

> In the automated option, I would change the default test from ADF to KPSS, since it's more standard and gives better results in my experience. Having this as a single class with regular fractional differentiation should be easy to use.

This could be a user choice - there is a type of object in `sktime`, the "parameter estimator", which includes stationarity tests.

We currently have 8 of these in the register, all from third party packages (`statsmodels` and `arch`):
https://www.sktime.net/en/latest/api_reference/param_est.html#stationarity-estimators

So one could set KPSS as a default, but let the user pass any of these.

> @fkiraly I don't know anything about the maintainer, I'm quite puzzled about this myself. There was also no notice about looking for maintainers, and nothing in the fracdiff issues.

The abruptness, lack of messaging, surprisal of the user base is similar to other higher-profile shutdowns I've seen recently, it is quite odd - though for the purpose of this issue it is not really useful to speculate.

The lessons we are drawing for `sktime` from these is that governance is important, and in particular risk mitigating operations on multiple levels (devops, legal, security).

> I'll try to find time to implement this, but I'm not sure how soon I can start. If anyone else would like to pick this up, feel free to do so.

For this to be a good first issue, there needs to be a concrete plan. I would say:

* move `fracdiff` - or just the files containing `fdiff` - to `sktime.libs.fracdiff`
* write `FractionalDifferencer` class

For ARFIMA, what would be the approach we should adopt? De-novo, or combining fracdiff and and an existing ARIMA?

**Comment by j-adamczyk** [comment]
@fkiraly

partial fork makes sense then, with just `fdiff`, and writing `FractionalDifferencer` using it.

For ARFIMA, I think that there should be 2 classes:
- `ARFIMA`, wrapping a selected ARIMA implementation (with set `d=0` in there) and set fractional `d`
- `AutoARFIMA`, similarly wrapping an AutoARIMA implementation, and automated fractional `d` selection

Should user be able to pass `d="auto"`, and have semi-automated model? I think it makes sense.

**Comment by fkiraly** [comment]
Quick question: would ARFIMA be different than just pipelining `FractionalDifferencer` with any ARIMA where d=0?

**Comment by j-adamczyk** [comment]
Yes, exactly. Do you think that it should stay separate then?

**Comment by j-adamczyk** [comment]
Makes sense, it can stay separately. Maybe we can just add information to `FractionalDifferencer` docs that it can be used to create ARFIMA models? I think that this name should appear somewhere in the docs, since this is relatively less known group of models.

**Comment by fkiraly** [comment]
> Makes sense, it can stay separately. Maybe we can just add information to `FractionalDifferencer` docs that it can be used to create ARFIMA models? I think that this name should appear somewhere in the docs, since this is relatively less known group of models.

Yes, makes sense - and people familiar with ARFIMA models may not be aware that they can get it from ARIMA and pipelining; for the exact design, how about a single ARFIMA, and users can select any of the ARIMA models as components?

**Comment by fkiraly** [comment]
Further comment about how I would deal with `FracdiffStat`.
This is internally using adf stationarity test from `statsmodels`, which exists in `sktime` as one of multiple stationarity tests

The steps I would carry out:
* I would move it to `BaseTransformer` from `sktime`
* replace the `fracdiff` class with the `sktime` stationarity test
* make the stationarity test an `__init__` argument - default being ADF or KPSS (tbd), so allow the user to choose any stationarity test

**Comment by fkiraly** [comment]
FYI, @dinobektesevic, I am planning to resolve the sklearn compatibility issue and wrap up vendoring of fracdiff.

Are you planning to work on the `BaseTransformer`-s? I think there are at least two to consider, `FractionalDifferencer`, and `FracdiffStat`.

An open question is where would get the inverse to `FractionalDifferencer` somewhere, this is crucial for use in pipelines, but does not seem to be supplied by the library or the sklearn transformer?

**Comment by fkiraly** [comment]
Sure, makes sense. I'll wrap up the interface compatibility then, that should be enough to help users to transition to the vendored lib which continues to be maintained. We can add the `sktime` transformations at your own pace if you're up for it.

**Comment by fkiraly** [comment]
reopening due to remaining tasks:

* `sktime` interface as transformatoins above
* ARFIMA
